### DISCUSSION TITLE
envconsul deprecation

### DISCUSSION THREAD

**Discussion by Ducky6944** [Homebrew/discussions#5564] [Tap maintenance and Homebrew development] [answered]
### Output of `brew config`

```shell
NA
```

### Output of `brew doctor`

```shell
NA
```

### Description of issue

I didn't want to setup a pull request for a formulae that was intentionally deprecated before I understood why. `envconsul` comes a precompiled binary or is installed via go. The deprecation message states that it's dependent on deprecated `consul.` It lets you interact with `consul` and `vault` but is not a dependency of.

**Comment by gromgit** [comment]
Putting together the cryptic clues you left, I assume you're talking about `fabio`.

I don't use any of the products you mention, but a quick source check reveals that there are at least two "points of contact" that `fabio` has with Consul:
1. [Importing](https://github.com/fabiolb/fabio/blob/642e425cf664e23d257ca357b7b1bfd1e432683d/cert/consul_source.go#L15) the [Consul API source](https://github.com/hashicorp/consul/tree/main/api) at build time. HashiCorp has [publicly committed to keeping all their APIs licensed under MPL-2.0](https://www.hashicorp.com/license-faq#aug-10-announcement), so this should not be an issue.
2. Testing against, and being configured by, a running Consul. This seems to be the real stumbling block, and not something easily worked around.

**Reply by Ducky6944** [reply]
It's not that cryptic. I am talking about `envconsul` as in `brew info envconsul`. It's a go package that has no dependencies, that I am aware of, on `consul`. Despite it's namesake it's just a tool you can use with consul and vault. 

The dependencies section states that go is the only dependency. 

This is not a huge issue though because hashicorp has a tap which includes it, but I don't think it should be deprecated.

```
> brew info envconsul
==> envconsul: stable 0.13.2 (bottled)
Launch process with environment variables from Consul and Vault                                  
https://github.com/hashicorp/envconsul                                                           
Deprecated because it depends on soon to be deprecated consul! It will be disabled on 2025-02-05.
Installed                                                                                        
/opt/homebrew/Cellar/envconsul/0.13.2 (7 files, 11.7MB) *                                        
  Poured from bottle using the formulae.brew.sh API on 2024-07-07 at 13:48:14                    
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/e/envconsul.rb                 
License: MPL-2.0                                                                                 
==> Dependencies                                                                                 
Build: go ✔                                                                                      
==> Analytics                                                                                    
install: 20 (30 days), 58 (90 days), 366 (365 days)                                              
install-on-request: 19 (30 days), 57 (90 days), 353 (365 days)                                   
build-error: 0 (30 days)
```

**Reply by gromgit** [reply]
> The dependencies section states that go is the only dependency.

`brew info` only lists runtime dependencies. The formula lists `consul` as a _test_ dependency, which is exactly the same as `fabio`'s issue 2 above. (Side note: `fabio` really should list `consul` as only a test dependency.)

**Reply by Ducky6944** [reply]
Gotcha. I didn't see that in the formula and brew info didn't list it. It seems like changing the test dependencies would be better than deprecating the formula as they are arbitrary in this case I think.

**Reply by gromgit** [reply]
Given that both `envconsul` and `fabio` are written specifically to interact with Consul, how would you change the test dependencies while performing actually-useful tests?

**Reply by Ducky6944** [reply]
I don't use consul at all. I don't have it installed. envconsul also works with vault with nothing specifically tying it to consul unless you want to use it with consul.

**Reply by gromgit** [reply]
Vault is not an option, since it's also deprecated for the same reason.

**Reply by Ducky6944** [reply]
Vault and Consul deprecation is due to a license change, which makes sense. envconsul is not changing licenses that I'm aware of. It seems like poor form to remove open source tools because of a test dependency. There is plenty of opensource software that supports proprietary software. How do those tests work?

A few questions:
Are test dependencies required in core? 
Is `envconsul --version` and a `sha` enough verification?
Is there a mechanism for grabbing precompiled packages in a brew formulae? If so they release precompiled packages, which I would assume negate the need to test. 

Is it strange that brew is testing software against other software and not against itself where possible for expected behavior? Whats the reasoning behind that? Envconsul is it's own program even if its developed to use in conjunction with other software.

**Reply by Ducky6944** [reply]
An example might be 'https://github.com/Homebrew/homebrew-core/blob/master/Formula/e/exercism.rb'. 
```ruby
assert_match version.to_s, shell_output("#{bin}/exercism version")
```

**Reply by gromgit** [reply, marked as answer]
> There is plenty of opensource software that supports proprietary software. How do those tests work?

Depending on the author, it ranges from "manually, and only if someone reports an issue" to a full GitHub-hosted CI environment like Homebrew.

> Are test dependencies required in core?

Yes if they're used for the test. Homebrew's CI starts with a clean slate every time, so _not_ declaring dependencies on stuff you use in the test will naturally result in a failed test.

> Is `envconsul --version` and a `sha` enough verification?

[Not under normal circumstances.](https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula) Exceptions _might_ be granted if you can convince the Homebrew maintainers otherwise, but "help" or "version" tests are generally considered insufficient.

> Is there a mechanism for grabbing precompiled packages in a brew formulae?

Sure. Many third-party formulae hosted by software authors have their `url` point to a binary release tarball, from which they install their desired assets.

But [that's not allowed in the core tap.](https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae)

> Is it strange that brew is testing software against other software and not against itself where possible for expected behavior? Whats the reasoning behind that? 

Tests are meant to catch issues with proper software functioning, otherwise they might as well not be run. The majority of tests in Homebrew formulae "tests against itself", simply because they don't need anything else to test proper functioning. Other formulae like `envconsul` and `pgcron` can't be usefully tested without their "host" software (Consul/Vault and PostgreSQL respectively); those dependencies would have to be installed by CI during the test phase.

> `assert_match version.to_s, shell_output("#{bin}/exercism version")`

I hadn't even started using Homebrew when the `exercism` formula was added, but since it's a client for a specific online service that can't be reliably replicated in a CI environment,[^1] this "test of limited utility" was allowed to stand.

[^1]: I also seem to recall plans to limit network access in Homebrew CI to specific phases for security reasons, and that the test phase is not included in that allowlist.

All that said, you _might_ be able to convince the Homebrew maintainers to look the other way and [downgrade the test's usefulness](https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula) (it was clearly done for `exercism` after all), but that starts with filing a PR with the desired changes and providing justification. Nothing happens until you do that.

**Reply by Ducky6944** [reply]
Genuinely appreciate the answers and insight into the these issues. I don't think this is a hill I want to die on, but this has opened my eyes regarding the decisions maintainers deal with and design of brew. Personally I'm still conflicted with this train of thought, but it's also completely reasonable to just utilize the wonderful tap system that exists.

Thank you!

**Reply by p-linnane** [reply]
Thank you @gromgit as usual for a well thought out and formatted response 💜

@typkrft If you open a PR with your proposed changes, we'll happily review it. Just speak to your rationale and link back to this discussion for context. No guarantees it will be undeprecated, but we'll at least see if we can make it work.
